### ISSUE TITLE
The result of inference using EncoderDecoderASR differs from transcribed text in wer_eval_clean.txt file.

### ISSUE THREAD

**Issue by starcell** [speechbrain/speechbrain#1590]
I trained model using recipe/KosponSpeech, and got the transcribed text as below.
I used speechbrain version 0.5.12.
.....
KsponSpeech_E00099, %WER 0.00 [ 0 / 4, 0 ins, 0 del, 0 sub ]
반말을 ; 할 ; 수가 ; 없어
= ; = ; = ; =
반말을 ; 할 ; 수가 ; 없어
.....

But I got different result using EncoderDecoderASR as below.

from speechbrain.pretrained import EncoderDecoderASR
asr_model = EncoderDecoderASR.from_hparams()

asr_model.transcribe_file('KsponSpeech_E00099.wav')
.....
'짯 반말을 할 수가 없어핼 수가 없어읎어읎어쯧'
.....

It is so different that transcribed from the same wave file.

The yaml file that I used in training is [conformer_medium.yaml](https://github.com/speechbrain/speechbrain/blob/develop/recipes/KsponSpeech/ASR/transformer/hparams/conformer_medium.yaml)

The yaml file that I used in inferencing is [hyperparams.yaml](https://huggingface.co/speechbrain/asr-conformer-transformerlm-ksponspeech/blob/main/hyperparams.yaml)

This module worked well with sb version 0.5.10. but doesn't work with sb version 0.5.12.
Where can I find the solution?

This issue is related with #1585.

**Comment by TParcollet** [comment]
@Adel-Moumen please pause the BeamSearch work and have a look at this, it looks bad and abnormal.

**Comment by starcell** [comment]
@TParcollet
Thank you for your reply.
Would you tell me more about the method in detail?
If you tell me how to pause the BeamSearch work, it will be very helpful.

**Comment by Adel-Moumen** [comment]
Hello, 

Thanks for raising this issue.

> This module worked well with sb version 0.5.10. but doesn't work with sb version 0.5.12.
So, if I understand correctly, the 0.5.10 version will give you the expected results with the `EncoderDecoderASR` while the 0.5.12 will give you incorrect results right?

**Comment by starcell** [comment]
Yes, right.
0.5.10 and 0.5.11 version gave me the expected results, but not with 0.5.12, and 0.5.13.

**Comment by AsuMagic** [comment]
On `0.5.10`/`0.5.11`, did you also train the model, or did you use a pre-trained one?  
Also, what interface did you use for inference? `EncoderDecoderASR` doesn't seem like it existed at the time as it was introduced by PR #534 (which should mean it's there since `0.5.12`).

Could you upload the model you have trained in `0.5.12`/`0.5.13`?  
I can't test it due to #1585 causing errors when using `speechbrain/asr-conformer-transformerlm-ksponspeech`.

**Comment by starcell** [comment]
@AsuMagic 
I trained in v0.5.13, my result is below.
'짯 반말을 할 수가 없어쿱 없어읎어 없어웍'
(correct transcription : 반말을 할 수가 없어)

I used EncoderDecoderASR.from_hparams() in inference.

When I trained in 0.5.11, I've got correct result.(반말을 할 수가 없어)

I uploaded my trained model files in v0.5.13 in the following link.

https://huggingface.co/npark/asr-conformer-ksponspeech

Thank you for your reply.

**Comment by starcell** [comment]
> 
@AsuMagic

PR #534  (which should mean it's there since 0.5.12).
--> But there is EncoderDecoderASR interface in v0.5.11.

https://github.com/speechbrain/speechbrain/blob/v0.5.11/speechbrain/pretrained/interfaces.py

**Comment by starcell** [comment]
I don't think this is similar to issue #1616. He said that inference of test data in Librispeech dataset were successful, but the example file was not infered correctly.
In my issue, all test data in KsponSpeech dataset evaluated properly, but the data are not infered correctly by using EncoderDecoderASR interface.

**Comment by AsuMagic** [comment]
> --> But there is EncoderDecoderASR interface in v0.5.11.

It seems I got lost in the git history, you're right :)  
This might make the issue easier to debug down the road.

> I don't think this is similar to issue #1616. He said that inference of test data in Librispeech dataset were successful, but the example file was not infered correctly.
> In my issue, all test data in KsponSpeech dataset evaluated properly, but the data are not infered correctly by using EncoderDecoderASR interface.

I don't think so either anymore. I was suspecting it despite this because some behavior was similar when I was testing, but it only was indicating that beam search was struggling, not that beam search itself is broken.

Anyway, after an embarrassing debugging session where I was testing the wrong file for one hour or two, I can reproduce your issue for `v0.5.12` and `v0.5.13` as you said.  
`v0.5.11` fails to load the model due to a parameter loading error (unrelated to the huggingface loading issue there is with that version, which you can work around by providing a local path instead of a HF repo path).  
I found that applying #1329 over `v0.5.11` makes it possible to infer using the model again. The same broken output happens for this audio file as with `v0.5.12`.  
I don't yet know what exactly this implies, though... (FWIW, not including the `TransformerASR` changes from that PR does not affect inference.)

**Comment by Adel-Moumen** [comment]
> @AsuMagic I found something.
> 
> Line 315 of https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/lobes/models/transformer/TransformerASR.py
> 
> In TransformerASR encode()
> 
> The below must be added before return,
> 
> ```
>     if self.attention_type == "RelPosMHAXL":
>         # FIXME we use pos embs also on enc output
>         encoder_out = encoder_out + self.positional_encoding_decoder(
>             encoder_out
>         )
> ```
> 
> Would you check and fix it?

Hello @TParcollet could you please have a look at that? It seems that @starcell is right. @AsuMagic told me that now the inference is exactly the same as the one expected when adding this block of code. Thanks.

**Comment by AsuMagic** [comment]
BTW, this can be used for reproduction (if whoever implements the PR wants to poke around):

```python
from speechbrain.pretrained import EncoderDecoderASR
asr_model = EncoderDecoderASR.from_hparams(source="npark/asr-conformer-ksponspeech", savedir="pretrained_models/asr-conformer-transformerlm-ksponspeech")
audio_path = ".npark/asr-conformer-ksponspeech/KsponSpeech_E00099.wav"
print(asr_model.transcribe_file(audio_path))
```

Output without the suggested patch (bad):  
`짯 반말을 할 수가 없어쿱 없어읎어 없어웍`

Output **with** the suggested patch (good):  
`반말을 할 수가 없어`

Paging in @anautsch since this seems pretty relevant to testing.  
tl;dr: It looks like a mismatch was introduced between `TransformerASR.forward` and `TransformerASR.encode` in #1329. The positional encoding must be added or the transcription is broken in the inference/`EncoderDecoderASR` code path.

**Comment by TParcollet** [comment]
@popcornell can you verify that this makes sense, it looks like it is related to something you know well.

**Comment by AsuMagic** [comment]
FWIW, I've checked the recipes for `RelPosMHAXL` users and there "only" seems to be three recipes actually using it and that should be directly affected:
- `recipes/LibriSpeech/ASR/transformer/hparams/conformer.yaml`
- `recipes/KsponSpeech/ASR/transformer/hparams/conformer_small.yaml`
- `recipes/KsponSpeech/ASR/transformer/hparams/conformer_medium.yaml`

**Comment by starcell** [comment]
>
There is mistake. 
recipes/LibriSpeech/ASR/transformer/hparams/conformer.yaml
recipes/LibriSpeech/ASR/transformer/hparams/conformer_small.yaml (not KsponSpeech)
recipes/KsponSpeech/ASR/transformer/hparams/conformer_medium.yaml

**Comment by popcornell** [comment]
@AsuMagic observations are correct IMO. 
Transformer code has lots of repetitions right now unfortunately and probably is better to refactor to reuse as much as possible stuff. Otherwise such stuff is more prone to happen ( and it was my fault this time..).

**Comment by TParcollet** [comment]
@AsuMagic interested in fixing? :D

**Comment by AsuMagic** [comment]
> @AsuMagic interested in fixing? :D

@TParcollet I'm going to be a bit too busy this month, so hopefully someone else can do it.

**Comment by popcornell** [comment]
https://github.com/speechbrain/speechbrain/blob/801b1501b0bde2a940fcb71af44b69b07eafb9f5/speechbrain/pretrained/interfaces.py#L634 

The problem I think it is because the `EncoderDecoderASR` object `transcribe` method does not anymore rely on the Transformer `decode` method. This bypasses the pos encoding code.

**Comment by TParcollet** [comment]
I am not sure to see what @popcornell is mentioning here. encode_batch is calling the forward function of TransformerASR (if properly defined in the yaml, as it is the case in this issue). TransformerASR forward function is doing positional encoding. I think the problem has been fixed by removing the pos embedding to the decoding by an above PR.

**Comment by popcornell** [comment]
Yes only tgt sequence now has pos encoding in the decoder.  So I think now it is fine ? https://github.com/speechbrain/speechbrain/blob/8fc31edc763e5b8860600ca806ff7c1575bc6aeb/speechbrain/lobes/models/transformer/TransformerASR.py#L396C20-L396C37

**Comment by Adel-Moumen** [comment]
Hello @starcell, 

I'd like to work on the issues related to Ksponspeech in the incoming weeks. However, I have some troubles to get access to the dataset. Do you mind if I ask you to share privately the dataset so that I can work on that? I will of course remove it when I will have fix every issues (for instance, one of them is to retrain our transducer on KsponSpeech since we made the transducer very competitive on librispeech etc)

**Comment by starcell** [comment]
> Hello @starcell,
> 
> I'd like to work on the issues related to Ksponspeech in the incoming weeks. However, I have some troubles to get access to the dataset. Do you mind if I ask you to share privately the dataset so that I can work on that? I will of course remove it when I will have fix every issues (for instance, one of them is to retrain our transducer on KsponSpeech since we made the transducer very competitive on librispeech etc)



> Hello @starcell,
> 
> I'd like to work on the issues related to Ksponspeech in the incoming weeks. However, I have some troubles to get access to the dataset. Do you mind if I ask you to share privately the dataset so that I can work on that? I will of course remove it when I will have fix every issues (for instance, one of them is to retrain our transducer on KsponSpeech since we made the transducer very competitive on librispeech etc)

**Comment by starcell** [comment]
Hello @Adel-Moumen,

I appreciate your interest in this issue and your efforts to experiment with Ksponspeech dataset for Korean speech recognition.

But, I'm very unhappy not to share the dataset with you.
I asked approval from the relevant authorities(NIA) to share the dataset with someone abroad. They answered it is not allowed to transfer the dataset overseas at any reason and at any means.

I sincerely apologize and please feel free to contact me anytime if you need any further help.

**Comment by Adel-Moumen** [comment]
Hello @ddwkim, hope you are doing well! I'm pinging you about the current state of ksponspeech. We currently have four open issues #1590, #2072, #1585, and #1528 related to this dataset. Unfortunately, I wasn't able to get my hands on the dataset. I'd like to ask you if you could investigate a bit more on this dataset and update the recipe to match our latest Conformer changes (see, for instance, LibriSpeech: https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/ASR/transformer). It would require updating the recipe to match what we did in LibriSpeech and to retrain the model and make sure that the inference is fine. Otherwise, we will need to remove this dataset from SpeechBrain since as core maintainers we cannot maintain it. 

One alternative could be to swap to Zeroth-Korean since it is publicly available on Open-SLR. Could you please get back to me asap so that we can move forward? Thank you very much.

**Comment by ddwkim** [comment]
Sure do. Thank you for reaching me out. I will update the recipe to meet with the latest updates, and share the dataset to allow maintainers to check the validity of the model, while minimizing the inconvenience may have incurred to our community.

In terms of Zeroth Korean, we considered using as our dataset for the conformer model, but its limited amount, which is less than 100 hours, hindered us from decision to adopt, as most fully neural network based E2E models heuristically require approximately 1000 hours-length of audio. This is the reason as to why the maintainers of Zeroth Korean only provided TDNN-chain model.

Another possibility is to train a model on massive Korean unlabeled datasets (or find one already trained in this way), and fine-tune it on Zeroth-Korean. I expect that this will allow the E2E ASR models to be trained and maintained without the aid on my side. That being said, collection of massive Korean datasets / or a model trained in this way will be a possible roadblock.

Nevertheless, I will fix the issues within a month, and get to you with solutions as soon as possible.

**Comment by Adel-Moumen** [comment]
Dear @ddwkim, 

> Sure do. Thank you for reaching me out. I will update the recipe to meet with the latest updates, and share the dataset to allow maintainers to check the validity of the model, while minimizing the inconvenience may have incurred to our community.

Thanks a lot but I need to warn you that sharing this dataset outside of Korea may put you in a difficult position since it is forbidden to do so (which is sad since as open-source maintainers we cannot widespread Korean ASR systems). If you want to share it, I agree to remove it or to keep it private and do not share it with anyone else.

> In terms of Zeroth Korean, we considered using as our dataset for the conformer model, but its limited amount, which is less than 100 hours, hindered us from decision to adopt, as most fully neural network based E2E models heuristically require approximately 1000 hours-length of audio. This is the reason as to why the maintainers of Zeroth Korean only provided TDNN-chain model.

Ok I understand. Thanks :)

> Another possibility is to train a model on massive Korean unlabeled datasets (or find one already trained in this way), and fine-tune it on Zeroth-Korean. I expect that this will allow the E2E ASR models to be trained and maintained without the aid on my side. That being said, collection of massive Korean datasets / or a model trained in this way will be a possible roadblock.

This could be a solution but this means a lot of extra engineering work and for most researchers, they would like to have a from-scratch ASR baseline (e.g., the conformer one) to compare their results with...

> As the inability to move the dataset abroad is being the issue at hand, I will also attempt to find a completely open-sourced way of building a Korean recipe. Such as researching another dataset and so on.
> I expect that it may take a month to meet with my schedule, but nevertheless get back to you with solutions as soon as possible.

Ok keep me posted and thanks again for your prompt reactivity. This is really great to have you part of the speechbrain community/developers :)

**Comment by Adel-Moumen** [comment]
Hi please see #3013.
